server/src/main/java/org/elasticsearch/cluster/coordination/LagDetector.java

```diff
@@ -0,0 +1,168 @@
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.common.util.concurrent.ConcurrentCollections.newConcurrentMap;
+
+/**
+ * A publication can succeed and complete before all nodes have applied the published state and acknowledged it; however we need every node
+ * eventually either to apply the published state (or a later state) or be removed from the cluster. This component achieves this by
+ * removing any lagging nodes from the cluster after a timeout.
+ */
+public class LagDetector {
+
+    private static final Logger logger = LogManager.getLogger(LagDetector.class);
+
+    // the timeout for each node to apply a value after the end of publication, before being removed from the cluster
+    public static final Setting<TimeValue> CLUSTER_FOLLOWER_LAG_TIMEOUT_SETTING =
+        Setting.timeSetting("cluster.follower_lag.timeout",
+            TimeValue.timeValueMillis(90000), TimeValue.timeValueMillis(1), Setting.Property.NodeScope);
+
+    private final TimeValue clusterStateApplicationTimeout;
+    private final Consumer<DiscoveryNode> onLagDetected;
+    private final Supplier<DiscoveryNode> localNodeSupplier;
+    private final ThreadPool threadPool;
+    private final Map<DiscoveryNode, NodeAppliedStateTracker> appliedStateTrackersByNode = newConcurrentMap();
+
+    public LagDetector(final Settings settings, final ThreadPool threadPool, final Consumer<DiscoveryNode> onLagDetected,
+                       final Supplier<DiscoveryNode> localNodeSupplier) {
+        this.threadPool = threadPool;
+        this.clusterStateApplicationTimeout = CLUSTER_FOLLOWER_LAG_TIMEOUT_SETTING.get(settings);
+        this.onLagDetected = onLagDetected;
+        this.localNodeSupplier = localNodeSupplier;
+    }
+
+    public void setTrackedNodes(final Iterable<DiscoveryNode> discoveryNodes) {
+        final Set<DiscoveryNode> discoveryNodeSet = new HashSet<>();
+        discoveryNodes.forEach(discoveryNodeSet::add);
+        discoveryNodeSet.remove(localNodeSupplier.get());
+        appliedStateTrackersByNode.keySet().retainAll(discoveryNodeSet);
+        discoveryNodeSet.forEach(node -> appliedStateTrackersByNode.putIfAbsent(node, new NodeAppliedStateTracker(node)));
+    }
+
+    public void clearTrackedNodes() {
+        appliedStateTrackersByNode.clear();
+    }
+
+    public void setAppliedVersion(final DiscoveryNode discoveryNode, final long appliedVersion) {
+        final NodeAppliedStateTracker nodeAppliedStateTracker = appliedStateTrackersByNode.get(discoveryNode);
+        if (nodeAppliedStateTracker == null) {
+            logger.trace("node {} applied version {} but this node's version is not being tracked", discoveryNode, appliedVersion);
```

Comment:
Mmm, ok, we're under the Coordinator mutex here. Changed to an assertion in cdcc190bd21a85b0a6e72f9c3942625bab63615a.

Reply:
This looks to be actually required because setAppliedVersion can be called for an earlier publication after we possibly have started another publication that has since removed the node.